### PR TITLE
Atomic values for the get_service_node_registration_cmd endpoint

### PR DESCRIPTION
**Pull request by darcys22**

### CODE REVIEW

File: src/rpc/core_rpc_server.cpp [hunk 1]
```diff
@@ -3008,26 +3008,44 @@ namespace cryptonote { namespace rpc {
...
-    uint64_t const curr_height   = m_core.get_current_blockchain_height();
-    uint64_t staking_requirement = service_nodes::get_staking_requirement(nettype(), curr_height);
+    std::optional<uint64_t> height = m_core.get_current_blockchain_height();
+    auto hf_version = get_network_version(nettype(), *height);
+    uint64_t staking_requirement = service_nodes::get_staking_requirement(nettype(), *height);
 
     {
       uint64_t portions_cut;
-      if (!service_nodes::get_portions_from_percent_str(req.operator_cut, portions_cut))
+      if (hf_version >= hf::hf19_reward_batching)
```
Comment (jagerman) [inline src/rpc/core_rpc_server.cpp:3017]: I don't think we need this `if` anymore; we can *always* assume ≥hf19 rules now.

(Actually, come to think of it, there's probably some delete-me-after-hf19 code we can remove as well).

File: src/rpc/core_rpc_server.cpp [hunk 2]
```diff
@@ -3008,26 +3008,44 @@ namespace cryptonote { namespace rpc {
...
+          return res;
+        }
       }
 
       args.push_back(std::to_string(portions_cut));
     }
 
     for (const auto& [address, amount] : req.contributions)
     {
+      if (hf_version >= hf::hf19_reward_batching)
```
Comment (jagerman) [inline src/rpc/core_rpc_server.cpp:3040]: Likewise

File: src/cryptonote_core/service_node_rules.cpp [hunk 3]
```diff
@@ -280,4 +280,18 @@ bool get_portions_from_percent_str(std::string cut_str, uint64_t& portions) {
   return false;
 }
 
+uint64_t get_atomic_amount_from_percent_str(std::string cut_str) {
+  const auto pct = parse_fee_percent(cut_str);
+  if (!pct)
+    throw "could not parse fee percent";
+
+  if(*pct < 0.0 || *pct > 100.0)
+    throw "fee percent out of bounds";
```
Comment (jagerman) [inline src/cryptonote_core/service_node_rules.cpp:289]: This needs an exception type (I'm sad that being able to throw a raw c string pointer even compiles in C++).

Comment (jagerman) [reply src/cryptonote_core/service_node_rules.cpp:289]: Probably `std::out_of_range` for this one, and `std::invalid_argument` for the first one.

File: src/cryptonote_core/service_node_rules.cpp [hunk 4]
```diff
@@ -280,4 +280,18 @@ bool get_portions_from_percent_str(std::string cut_str, uint64_t& portions) {
...
+
+  auto atomic_amount = static_cast<uint16_t>(std::lround(*pct / 100.0 * cryptonote::STAKING_FEE_BASIS));
+  if (*pct == 100.0)
+    atomic_amount = cryptonote::STAKING_FEE_BASIS;
+
+  return atomic_amount;
+}
```
Comment (jagerman) [inline src/cryptonote_core/service_node_rules.cpp:296]: `rpc_command_executor.cpp` should get updated to call this function (it basically has duplication of this logic, which I'm thinking is where it came from).

File: src/cryptonote_core/service_node_rules.cpp [hunk 5]
```diff
@@ -280,4 +280,18 @@ bool get_portions_from_percent_str(std::string cut_str, uint64_t& portions) {
   return false;
 }
 
+uint64_t get_atomic_amount_from_percent_str(std::string cut_str) {
```
Comment (jagerman) [inline src/cryptonote_core/service_node_rules.cpp:283]: This function should probably be renamed; HF19 registrations are 0-10000 integer values, so I guess they are "atomic", but don't want to confuse it with atomic oxen.  These is, I believe, coincidentally the same as financial basis points, so perhaps "percent_to_basis_points" would work?

File: src/cryptonote_core/service_node_rules.cpp [hunk 6]
```diff
@@ -280,4 +280,18 @@ bool get_portions_from_percent_str(std::string cut_str, uint64_t& portions) {
   return false;
```
Comment (jagerman) [inline src/cryptonote_core/service_node_rules.cpp:280]: I'm pretty sure this function won't be used anymore (after deleting the pre-hf19 registration code) so we should nuke it.

File: src/rpc/core_rpc_server.cpp (hunk 1 above)
Comment (jagerman) [reply src/rpc/core_rpc_server.cpp:3017]: I've just pushed code to remove pre-hf19 regs from `oxend prepare_registration`

File: src/cryptonote_core/service_node_rules.cpp (hunk 3 above)
Comment (ferchosnorlax7-dev) [reply src/cryptonote_core/service_node_rules.cpp]: Menú de navegación
[núcleo de bueyes](https://github.com/oxen-io/oxen-core)

[Código](https://github.com/oxen-io/oxen-core)
Asuntos
115
Fusionado
[Valores atómicos para el punto final get_service_node_registration_cmd](https://github.com/oxen-io/oxen-core/pull/1591#top)
#1591
[Jagerman](https://github.com/jagerman) fusionó 3 confirmaciones en [bueyes-io:desarrollador](https://github.com/oxen-io/oxen-core/tree/dev) de [darcys22:comando atómico-sn-reg-cmd](https://github.com/darcys22/loki-core/tree/atomic-sn-reg-cmd)  el 30 de agosto de 2022
[Conversación 8](https://github.com/oxen-io/oxen-core/pull/1591)
[Se compromete 3](https://github.com/oxen-io/oxen-core/pull/1591/commits)
[Cheques 0](https://github.com/oxen-io/oxen-core/pull/1591/checks)
[Archivos modificados 5](https://github.com/oxen-io/oxen-core/pull/1591/files)
Conversación
@[darcys22](https://github.com/darcys22)
@darcys22
darcys22
comentó
[el 30 de agosto de 2022](https://github.com/oxen-io/oxen-core/pull/1591#issue-1355227017)
No se proporcionó descripción.

@[darcys22](https://github.com/darcys22) darcys22 solicitó una reseña a [jagerman](https://github.com/jagerman) [Hace 3 años](https://github.com/oxen-io/oxen-core/pull/1591#event-7284241856)
Jägerman
[Jagerman](https://github.com/jagerman) revisado [el 30 de agosto de 2022](https://github.com/oxen-io/oxen-core/pull/1591#pullrequestreview-1091022763)
[src/rpc/core_rpc_server.cpp](https://github.com/oxen-io/oxen-core/pull/1591/files#diff-2213c2e8209890a49824ecb03acc9df9e2b666a52c0e9edf34d2bad7484a51c6)
Anticuado
Jägerman
[Jagerman](https://github.com/jagerman) revisado [el 30 de agosto de 2022](https://github.com/oxen-io/oxen-core/pull/1591#pullrequestreview-1091023001)
[src/rpc/core_rpc_server.cpp](https://github.com/oxen-io/oxen-core/pull/1591/files#diff-2213c2e8209890a49824ecb03acc9df9e2b666a52c0e9edf34d2bad7484a51c6)
Anticuado
Jägerman
[Jagerman](https://github.com/jagerman) revisado [el 30 de agosto de 2022](https://github.com/oxen-io/oxen-core/pull/1591#pullrequestreview-1091024651)
[src/cryptonote_core/reglas_del_nodo_de_servicio.cpp](https://github.com/oxen-io/oxen-core/pull/1591/files#diff-28f38fecc313f52a2c8d77e97825a2dc6412115f13757795dc48ea5bdf13c96e)
Anticuado
@jagerman
[Jägerman](https://github.com/jagerman)
el 30 de agosto de 2022
Esto necesita un tipo de excepción (me entristece que poder lanzar un puntero de cadena C sin formato incluso se compile en C++).

@jagerman
[Jägerman](https://github.com/jagerman)
el 30 de agosto de 2022
Probablemente std::out_of_rangepara éste y std::invalid_argumentpara el primero.

Comentario
 
Deja un comentario
No admitimos ese tipo de archivo.

Inténtalo de nuevo con GIF, JPEG, JPG, MOV, MP4, PNG, SVG, WEBM, WEBP, BMP, C, COPILOTMD, CPP, CPUPROFILE, CS, CSS, CSV, DEBUG, DMP, DOC, DOCX, DRAWIO, EML, FODG, FODP, FODS, FODT, GZ, HTM, HTML, IPYNB, JAVA, JS, JSON, JSONC, LOG, MD, MP3, MSG, ODF, ODG, ODP, ODS, ODT, PATCH, PDB, PDF, PHP, PPTX, PY, RTF, SH, SQL, TGZ, TIF, TIFF, TS, TSV, TSX, TXT, WAV, XLS, XLSM, XLSX, XML, YAML, YML o ZIP.

**Pull request merged**